### PR TITLE
Ensure we give a better error message if llvm-config is not in path.

### PR DESCRIPTION
**Pull request by stonea**
This is basically a redo of: https://github.com/chapel-lang/chapel/pull/18179

Which I had to back out due to failing smoke tests in: https://github.com/chapel-lang/chapel/pull/18191

I can reproduce the bug that was encountered in the smoke test by unsetting `CHPL_LLVM` and running `printchplenv`. I find this a little confusing since the name of the smoke test configuration would seem to indicate that it sets `CHPL_LLVM=none` (see here: https://chapel-ci.us.cray.com/job/chapel-code-smoke-test/2157/extra_config=none,label_exp=chapvm/) and if I explicitly set `CHPL_LLVM` to none I am not able to reproduce the issue.

Nevertheless, I want to proceed under the assumption that this was the bug.

--

Given that this failed smoke tests before, let me take the time to write a little bit about what I think the bug\fix for the smoke test failure is.

Before any of my changes (in the old PR or this one). We have a piece of Python code that is supposed to error out if `CHPL_LLVM` is system and no actual system path can be found here:

```python
def validate_llvm_config():
    llvm_val = get()
    llvm_config = get_llvm_config()
    if llvm_val == 'system':
        if llvm_config == 'none':
            error("CHPL_LLVM=system but could not find an installed LLVM"
                  " with one of the supported versions: {0}".format(
                  llvm_versions_string()))
```

The original issue motivating this issue is that if we set `CHPL_LLVM=system` and we don't actually have LLVM installed we get a bad error message that doesn't hint as to what the actual issue is.

To resolve this, the original PR (as well as this one) ensures that the `validate_llvm_config` function gets called before any other code that needs to call out to `llvm-config` is evaluated.  The second issue (addressed by the old PM as well as this one) is that `llvm_config` as it is returned by the `get_llvm_config()` function returns an empty string if we can't find a valid llvm-config. 

To make the error check work, in the previous PR I modified find_system_llvm_config() return 'none' instead of this empty string. The problem is another function, `has_compatible_installed_llvm` uses the value like this:

```python
def has_compatible_installed_llvm():
    llvm_config = find_system_llvm_config()
    if llvm_config:
        return True
    else:
        return False
```

Apparently if on an empty string in Python will return false, so the fact that I had it return 'none' instead of the empty string would cause `has_compatible_installed_llvm` to have the wrong behavior.

In this new PR I don't modify `find_system_llvm_config()` to return "none" (I keep at as the empty string), but I do add a check for the empty string back in `validate_llvm_config`.

### CODE REVIEW

File: util/chplenv/chpl_llvm.py [hunk 1]
```diff
@@ -163,11 +163,14 @@ def get_llvm_config():
...
+def validate_llvm_config(llvm_config=None):
     llvm_val = get()
-    llvm_config = get_llvm_config()
+    # We pass in llvm_config if has already been computed (so we don't
+    # end up in an infinite loop).
+    if llvm_config is None:
+      llvm_config = get_llvm_config()
     if llvm_val == 'system':
-        if llvm_config == 'none':
+        if not llvm_config or llvm_config == 'none':
```
Comment (daviditen) [inline util/chplenv/chpl_llvm.py:173]: I think this line would be more clear as `if llvm_config == '' or llvm_config == 'none':`.

**Review by daviditen — APPROVED** [review]

File: util/chplenv/chpl_llvm.py (hunk 1 above)
Comment (stonea) [reply util/chplenv/chpl_llvm.py]: Sure, updated in 4ca1e4dc8441d3bbc87ffde412129a4bccfaf0a0

**Comment by bradcray** [comment]
Andy: The row labels in that Matrix aren't particularly clear to me since `gasnet` is a CHPL_COMM setting and `llvm` is either a `CHPL_LLVM` or `CHPL_TARGET_COMPILER` setting (so they're not orthogonal, and `none` means different things for each of them).  I vaguely remember that we only added the LLVM row recently (though I can't recall who did), so would guess that `none` historically meant `CHPL_COMM=none`.  But I'm not sure what the implication would be for the CHPL_LLVM setting.  My guess would be that it would default to `system` if a reasonable version was available and `unset` otherwise.

Tagging @Maxrimus to see if he knows more about the above questions.

This also relates to the outstanding question we have about "Which configurations used to explicitly be testing LLVM that are now redundant in an 'LLVM-by-default' world?" and "Are we doing an appropriate amount of C back-end testing these days given that almost everything used to test the C back-end?"  (which I think of as "Where are we currently exposed w.r.t. not doing as much testing as we used to going into the release?")

**Comment by Maxrimus** [comment]
Ok so the llvm and gasnet configs were added back in 2017 by Tony Wallace:
https://stash.us.cray.com/projects/CHAPEL/repos/chapel-ci-config/commits/65053a8154d056d836d9b58e3258907e92de93ff
Although the llvm test was later tweaked by Michael to use a system llvm: https://stash.us.cray.com/projects/CHAPEL/repos/chapel-ci-config/commits/0ed39411635b30e6026aee16eb537ce678301fbf#config/meta.yaml
Besides that though, these tests have not been touched for years

I was actually just talking with @daviditen about these so he might have broader insight but from my understanding these tests are to just smoketest the default, the default with gasnet, and the default with llvm. So I'm not even sure we need the llvm row at this point, considering it is part of the default. It might even be better to swap that row to a C compiler?

**Comment by bradcray** [comment]
> the llvm and gasnet configs were added back in 2017

Oops, I was clearly mixed up with one of the other matrix configurations... probably the XC matrix where we test different target compilers (?).

If the `none` and `llvm` rows are testing the same configuration, I agree that they're redundant (well, the llvm row is because the `none` tests all three columns where `llvm` only tests the first).  And with a super-quick check it appears that they could be.  I:
* clicked into the first column configuration of each (https://chapel-ci.us.cray.com/job/chapel-code-smoke-test/extra_config=none,label_exp=chapvm/2190/console and https://chapel-ci.us.cray.com/job/chapel-code-smoke-test/extra_config=llvm,label_exp=chapvm/2190/console
* clicked on "Console output"
* searched for '[Debug] printchplenv'
* compared the settings

and it seems that the only differences are whether `CHPL_TARGET_COMPILER: llvm` and `CHPL_LLVM: system` are set or inferred (indicated by the `*`).

So unless something changes them after that point in the scripting (which seems unlikely, but...) then it seems they're doing the same thing.

This is a good example of the kind of semi-inadvertent redundancy in testing we were discussing today, and I agree that  where replacing the `llvm` row with a `CHPL_TARGET_COMPILER=gnu` smoke test would provide more coverage and remove the redundancy (where it makes sense that the LLVM row would test all three columns but gcc just the first given llvm's primacy now).

**Pull request merged**